### PR TITLE
[MHLO] add BroadcastOp constant folding and identity folding

### PR DESCRIPTION
**Pull request by qingyunqu**
* add BroadcastOp constant folding and identity folding

### CODE REVIEW

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc [hunk 1]
```diff
@@ -1351,6 +1351,27 @@ static LogicalResult Verify(BroadcastOp op) {
   return success();
 }
 
+OpFoldResult BroadcastOp::fold(ArrayRef<Attribute> attrs) {
+  auto type = getType().cast<RankedTensorType>();  
+  if (type == getOperand().getType()) {
+    auto sizesSize = broadcast_sizes().getType().getNumElements();
+    if (sizesSize != 0) {
```
Comment (joker-eph) [inline tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc:1358]: It's not clear to me what you're checking for here? Which test does this correspond to?

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc [hunk 2]
```diff
@@ -1351,6 +1351,27 @@ static LogicalResult Verify(BroadcastOp op) {
...
+    }
+    return getOperand();
+  }
+
+  // Constant fold when an operand is a splat tensor attribute.
+  if (!attrs[0] || !type.hasStaticShape()) return {};
+  auto splatOperandAttr = attrs[0].dyn_cast<SplatElementsAttr>();
+  if (!splatOperandAttr) return {};
+  // MLIR core bug (https://bugs.llvm.org/show_bug.cgi?id=46588): dense element
+  // attribute iterator not implemented for complex element types.
```
Comment (joker-eph) [inline tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc:1369]: This is fixed already, can you try enabling this and adding a test for it?

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc [hunk 3]
```diff
@@ -1351,6 +1351,27 @@ static LogicalResult Verify(BroadcastOp op) {
...
+    if (sizesSize != 0) {
+      return {};
+    }
+    return getOperand();
+  }
+
+  // Constant fold when an operand is a splat tensor attribute.
+  if (!attrs[0] || !type.hasStaticShape()) return {};
+  auto splatOperandAttr = attrs[0].dyn_cast<SplatElementsAttr>();
+  if (!splatOperandAttr) return {};
```
Comment (joker-eph) [inline tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc:1367]: Can you add test for these three conditions?

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc (hunk 1 above)
Comment (qingyunqu) [reply tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc:1358]: I think `if (type == getOperand().getType())` is enough. I will remove this checking.

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc (hunk 2 above)
Comment (qingyunqu) [reply tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc:1369]: There is an assertion failed when adding complex type test: 
`mlir/lib/IR/BuiltinAttributes.cpp:688: static mlir::DenseElementsAttr mlir::DenseElementsAttr::get(mlir::ShapedType, ArrayRef<mlir::Attribute>): Assertion `attr.isa<StringAttr>() && "expected string value for non integer/index/float element"' failed.`

Comment (qingyunqu) [reply tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc:1369]: So I think this should be fixed after fixing `DenseElementsAttr` in MLIR

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc (hunk 3 above)
Comment (qingyunqu) [reply tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc:1363]: done

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc (hunk 2 above)
Comment (joker-eph) [reply tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc:1365]: This is just not the right API to use to create a Dense attribute with Complex.
Instead there are multiple methods for it:

```
   // works with T = std::complex<U>
  static DenseElementsAttr get(const ShapedType &type, ArrayRef<T> values) {

  /// Constructs a dense float elements attribute from an array of APFloat
  /// values. Each APFloat value is expected to have the same bitwidth as the
  /// element type of 'type'. 'type' must be a vector or tensor with static
  /// shape.
  static DenseElementsAttr get(ShapedType type, ArrayRef<APFloat> values);

  /// Constructs a dense complex elements attribute from an array of APFloat
  /// values. Each APFloat value is expected to have the same bitwidth as the
  /// element type of 'type'. 'type' must be a vector or tensor with static
  /// shape.
  static DenseElementsAttr get(ShapedType type,
                               ArrayRef<std::complex<APFloat>> values);
```

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc (hunk 3 above)
Comment (joker-eph) [reply tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc:1363]: I think you're missing a case for the static shape test?

Comment (qingyunqu) [reply tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc:1364]: I can't create a constant value with dynamic shape, so I add dynamic shape identity test case instead.

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc (hunk 2 above)
Comment (qingyunqu) [reply tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc]: Done, thanks! I also fix the same problem in `BroadcastInDimOp`.

**Comment by qingyunqu** [comment]
@joker-eph  please review again. Thanks!

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc (hunk 3 above)
Comment (joker-eph) [reply tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc:1364]: Yes that's what I was looking for, thanks :)

**Review by joker-eph — APPROVED** [review]

**Pull request merged**